code/notificationbuilder/src/main/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/MultiIconNotificationBuilder.kt

```diff
@@ -0,0 +1,131 @@
+        )
+
+        val packageName = context.packageName
+        val notificationLayout = RemoteViews(packageName, R.layout.push_template_multi_icon)
+
+        val notificationManager =
+            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+        val channelIdToUse: String = notificationManager.createNotificationChannelIfRequired(
+            context,
+            pushTemplate
+        )
+
+        populateIconsForMultiIconTemplate(
+            context,
+            trackerActivityClass,
+            notificationLayout,
+            pushTemplate,
+            pushTemplate.templateItemList,
+            packageName
+        )
+
+        setCancelIcon(
+            notificationLayout,
+            pushTemplate,
+            packageName)
+
+        notificationLayout.setRemoteViewClickAction(
+            context,
+            trackerActivityClass,
+            R.id.five_icon_close_button,
+            null,
+            null,
+            pushTemplate.tag,
+            false
+        )
+
+        return AEPPushNotificationBuilder.construct(
+            context,
+            pushTemplate,
+            channelIdToUse,
+            trackerActivityClass,
+            notificationLayout,
+            notificationLayout,
+            R.id.carousel_container_layout
+        )
+    }
+
+    private fun setCancelIcon(
+        notificationLayout: RemoteViews,
+        pushTemplate: MultiIconPushTemplate,
+        packageName: String?
+    ) {
+        val iconString = pushTemplate.cancelIcon
+        notificationLayout.setRemoteViewImage(iconString, R.id.five_icon_close_button)
+    }
+
+    private fun populateIconsForMultiIconTemplate(
+        context: Context,
+        trackerActivityClass: Class<out Activity>?,
+        notificationLayout: RemoteViews,
+        pushTemplate: MultiIconPushTemplate,
+        items: MutableList<MultiIconPushTemplate.MultiIconTemplateItem>,
+        packageName: String?
+    ) {
+        var validImagesAddedCount = 0
+        for (item in items) {
+            val imageUri: String = item.iconUrl
+            val iconItem = RemoteViews(packageName, R.layout.multi_icon_template_item)
+            if (iconItem.setRemoteViewImage(imageUri, R.id.icon_item_image_view)) {
+                validImagesAddedCount++
+            }
+
+            trackerActivityClass?.let {
+                val interactionUri = item.actionUri ?: pushTemplate.actionUri
+                iconItem.setRemoteViewClickAction(
+                    context,
+                    trackerActivityClass,
+                    R.id.icon_item_image_view,
+                    interactionUri,
+                    null,
+                    pushTemplate.tag,
+                    pushTemplate.isNotificationSticky ?: false
+                )
+            }
+            notificationLayout.addView(R.id.icons_layout_linear, iconItem)
+        }
+        if (validImagesAddedCount < 3) {
```

Comment:
can we add the minimum valid images count to the constants to prevent a magic number warning?

Reply:
Done